### PR TITLE
fix: Ad-Hoc filters not applied because variables are not recognized by the AST parser

### LINKED ISSUE #714: Ad-hoc does not work, failing grafana template variables parsing
Ad-hoc filter fails to work in queries with grafana's variables due to parse failing.

**What happened**:
- Added ad-hoc variable to dashboard with panels with clickhouse datasource .
- Selected a column to filter by with value

**What you expected to happen**:
- See the data filter by ad-hoc filter

**How to reproduce it (as minimally and precisely as possible)**:
1 Add a clickhouse datasource
2 Create a dashboard with clickhouse datasource panel
3 Use any grafana's template in query, like SQL comment `/* ${__dashboard} */` 
4 Add variable ad-hoc with that datasource
5 Open JS console in browser
6 Select column from ad-hoc variable and you will see the error message like on this screenshot 
305916863-d4696b85-679a-4d84-8d59-cb8a5dbb0ec3

**Environment**:

- Grafana version: 10.3.3
- Plugin version: 4.0.2
- User OS & Browser: os x sonoma, chrome 120.0.6099.129

### PR DESCRIPTION
**Pull request by SpencerTorres**
Fixes #714 

Changes:
- Fixed the AST from not recognizing variables such as `${var}` `${var.key}` and `${var:singlequote}`. This was causing Ad-Hoc filters to not be applied, and prevented switching back to the Query Builder.
- Switching from SQL Editor back to Query Builder works again. This worked before, but was also broken when using variables. Also it would open an empty query, but now the AST is able to recognize the v4 builder options a little better.
- Fixed an issue where an empty database name would put `FROM "undefined"."table"`
- Multi filters (such as ones using `WHERE .. IN`) now allows for spaces. v3 allowed for `a,b,c` or `a, b, c`, v4 had just the first one but now it supports both. You can also include expressions and it works nicely, for example: `a, b, toString('c')`
- Added new unit tests that specifically focus the AST when using macros and variables.
- Re-added some tests for the AST + SQL generator. These were removed before because the SQL generator was separated from the utils module, but I've added it back. The original intent of these tests seems to be testing consistency when converting sql->builderOptions->sql.
- updated changelog

### CODE REVIEW

File: src/data/ast.ts [hunk 1]
```diff
@@ -1,31 +1,69 @@
...
     }
     sql = sql.substring(0, si) + replacementName + sql.substring(si + replaceFuncs[i].name.length);
   }
 
+  return [replaceFuncs, sql];
+}
+
+/**
+ * Replaces Grafana variables such as ${var} ${var.key} ${var.key:singlequote}
+ */
```
Comment (mshustov) [inline src/data/ast.ts:40]: let's add a link to appropriate section https://grafana.com/docs/grafana/latest/dashboards/variables/

File: src/data/ast.ts [hunk 2]
```diff
@@ -1,31 +1,69 @@
...
+
+  let regExpArray: RegExpExecArray | null;
+  while ((regExpArray = variableRegex.exec(sql)) !== null) {
+    replaceVariables.push({ startIndex: regExpArray.index, name: regExpArray[0], replacementName: '' });
+  }
+
+  // need to process in reverse so starting positions aren't affected by replacing other things
+  for (let i = replaceVariables.length - 1; i >= 0; i--) {
+    const si = replaceVariables[i].startIndex;
+    const replacementName = 'v' + (Math.random() + 1).toString(36).substring(7);
```
Comment (mshustov) [inline src/data/ast.ts:53]: nit: let's extract in a separate function or use uuid

File: src/data/ast.ts [hunk 3]
```diff
@@ -1,31 +1,69 @@
...
+  }
+
+  return [replaceVariables, sql];
+}
+
+export function sqlToStatement(sql: string): Statement {
+  const [replaceVars, variableSql] = replaceMacroVariables(sql);
+  const [replaceFuncs, macroSql] = replaceMacroFunctions(variableSql);
+  const replaceParts = replaceVars.concat(replaceFuncs);
+  sql = macroSql;
```
Comment (mshustov) [inline src/data/ast.ts:65]: let's avoid overriding variables. I'd rather rename the function signature
```typescript
export function sqlToStatement(rawSql: string): Statement {
  const [replaceVars, variableSql] = replaceMacroVariables(sql);
  const [replaceFuncs, sql] = replaceMacroFunctions(variableSql);
```

File: src/data/ast.test.ts [hunk 4]
```diff
@@ -15,5 +15,30 @@ describe('ast', () => {
       // this is formatted like this to match how pgsql generates its sql
       expect(toSql.statement(stm)).toEqual('SELECT (count (*) )  FROM foo');
     });
+
+    // https://github.com/grafana/clickhouse-datasource/issues/714
+    it('does not error when brackets/macros/variables are present', () => {
```
Comment (mshustov) [inline src/data/ast.test.ts:20]: Let's add more tests in a follow-up. I'd add more built-in ClickHouse functions, [query params](https://clickhouse.com/docs/en/interfaces/cli#cli-queries-with-parameters)

File: src/data/ast.test.ts [hunk 5]
```diff
@@ -15,5 +15,30 @@ describe('ast', () => {
...
+      const errLog = jest.spyOn(console, 'error');
+      const sql = `
+        SELECT
```
Comment (mshustov) [inline src/data/ast.test.ts:23]:
```suggestion
        /* ${__dashboard} */
        SELECT
```

File: src/data/sqlGenerator.ts [hunk 6]
```diff
@@ -515,12 +515,12 @@ const getColumnIdentifier = (col: SelectedColumn): string => {
...
 
 const getTableIdentifier = (database: string, table: string): string => {
-  const sep = (database === '' || table === '') ? '' : '.';
+  const sep = (!database || !table) ? '' : '.';
   return `${escapeIdentifier(database)}${sep}${escapeIdentifier(table)}`;
 }
 
 const escapeIdentifier = (id: string): string => {
-  return id === '' ? '' : `"${id}"`;
+  return id ? `"${id}"` : '';
```
Comment (mshustov) [inline src/data/sqlGenerator.ts:523]: let's add a few test as well. can be done in follow-up

**Review by asimpson — DISMISSED** [review]
Seems to work well, thanks! :trophy:

**Review by asimpson — APPROVED** [review]

**Pull request merged**